Adapt time-slice selection in High Contrast Mode.

nchevobbe opened the pull request:
Make sure the start and end grips look like button, make the
outbound of the selection even dimmer and add border to make
the selection super obvious.

Using registered property we can get the rgb component from
given values that we can then turn into SVG filters that
we can then apply on top of the icon we use, in CSS.
The time-slice zoom icon takes advantage of this new capability
to adapt to High Contrast Mode.

---

| | dark | light |
| --- | --- | --- |
| timeslice selected, end "grip" is hovered  | ![image](https://github.com/user-attachments/assets/556f7aa2-49e1-4424-ae2a-b8c399aab2bb)  | ![image](https://github.com/user-attachments/assets/0db90ff3-7f95-416a-8042-b9e9afcad2b1)   |
| hovered zoom icon  | ![image](https://github.com/user-attachments/assets/379645b3-1efa-465c-a521-a998a946dc7f)  |  ![image](https://github.com/user-attachments/assets/69739612-c2ae-473a-8641-1c7ed8f86439)  |
| active zoom icon  | ![image](https://github.com/user-attachments/assets/60fd6f38-7379-426e-a0a7-eb15cff9107f)  |  ![image](https://github.com/user-attachments/assets/74e75f45-8722-4815-8b17-fece85ebc0a7)  |

nchevobbe commented:
I'm curious about your opinion on the second commit to have more control hover the icon colors

mstange commented on src/components/timeline/Selection.css:
```diff
@@ -145,14 +150,21 @@
+.timelineSelectionOverlayZoomButton::before {
+  position: absolute;
+  align-content: center;
+  content: url(http://localhost:4242/20532f8cde285757f78c.svg);
```
`localhost:4242` :)

nchevobbe replied:
ah yes, will fix

mstange commented:
I think the filter solution is fine but you could use `<feFlood>` and `<feComposite>` so that you don't have to parse the rgba syntax. Example:

```svg
<filter id="--button-icon-color" color-interpolation-filters="sRGB">
  <feFlood flood-color="red"/>
  <feComposite operator="in" in2="SourceAlpha"/>
</filter>
```

mstange commented:
Oh, I found an even better solution! You can use `FillPaint` in combination with the `fill` property! I didn't realize this worked for HTML elements, too.

```html
<filter id="--use-fill-color" color-interpolation-filters="sRGB">
  <feComposite operator="in" in="FillPaint" in2="SourceAlpha"/>
</filter>
```

```css
fill: green;
filter: url(#--use-fill-color);
```

nchevobbe commented:
> Oh, I found an even better solution! You can use `FillPaint` in combination with the `fill` property! I didn't realize this worked for HTML elements, too.
> 
> ```
> <filter id="--use-fill-color" color-interpolation-filters="sRGB">
>   <feComposite operator="in" in="FillPaint" in2="SourceAlpha"/>
> </filter>
> ```
> 
> ```css
> fill: green;
> filter: url(#--use-fill-color);
> ```

Unfortunately this doesn't seem to work in Chromium browser. Would have been nice!

> I think the filter solution is fine but you could use `<feFlood>` and `<feComposite>` so that you don't have to parse the rgba syntax. Example:
> 
> ```
> <filter id="--button-icon-color" color-interpolation-filters="sRGB">
>   <feFlood flood-color="red"/>
>   <feComposite operator="in" in2="SourceAlpha"/>
> </filter>
> ```

Ah yes that's better, I'll do that

nchevobbe commented:
I don't really know flow, so not sure what's needed to please it here

julienw reviewed (approved):
I'm very newbie in terms of SVG filters. I looked at MDN but still I'm not sure I fully understand how filters work, but I'm not opposed.

I was a bit afraid that `@property` doesn't work in the old ESR, can you double check that the button still works as expected in that version? I don't mind that the forced-colors version doesn't work properly there, but at least the main way of working should work.

Thanks!

nchevobbe commented:
> I'm very newbie in terms of SVG filters. I looked at MDN but still I'm not sure I fully understand how filters work, but I'm not opposed.
> 
> I was a bit afraid that `@property` doesn't work in the old ESR, can you double check that the button still works as expected in that version? I don't mind that the forced-colors version doesn't work properly there, but at least the main way of working should work.
> 
> Thanks!

You mean current ESR (128) or the previous one (115) ? `@property` was added to 128, not sure how things look in 115 though

julienw commented:
> > I'm very newbie in terms of SVG filters. I looked at MDN but still I'm not sure I fully understand how filters work, but I'm not opposed.
> > I was a bit afraid that `@property` doesn't work in the old ESR, can you double check that the button still works as expected in that version? I don't mind that the forced-colors version doesn't work properly there, but at least the main way of working should work.
> > Thanks!
> 
> You mean current ESR (128) or the previous one (115) ? `@property` was added to 128, not sure how things look in 115 though

yeah I meant v115, [as far as I understand it's still supported](https://whattrainisitnow.com/release/?version=esr).

julienw commented on src/index.js:
```diff
@@ -44,6 +44,49 @@ window.geckoProfilerPromise = new Promise(function (resolve) {
+  };
+  defineSvgFiltersForColors();
+
+  const forcedColorsMql = window.matchMedia('(forced-colors: active)');
+  const darkSchemeMql = window.matchMedia('(prefers-color-scheme: dark)');
```
To fix the flow error happening here, you just need to add the right signature to `matchMedia` in https://github.com/firefox-devtools/profiler/blob/52f31449a52670689025a41b6477a5a04b9f2a43/src/types/globals/Window.js#L28

nchevobbe commented on src/index.js:
```diff
@@ -44,6 +44,49 @@ window.geckoProfilerPromise = new Promise(function (resolve) {
+      let filterEl = svgFiltersElement.getElementById(cssVariable);
+      let feFloodEl;
+      if (!filterEl) {
+        const xmlns = 'http://www.w3.org/2000/svg';
+        filterEl = document.createElementNS(xmlns, 'filter');
+        filterEl.id = cssVariable;
+        filterEl.setAttribute('color-interpolation-filters', 'sRGB');
+
+        feFloodEl = document.createElementNS(xmlns, 'feFlood');
+        const feCompositeEl = document.createElementNS(xmlns, 'feComposite');
+        feCompositeEl.setAttribute('operator', 'in');
+        feCompositeEl.setAttribute('in2', 'SourceAlpha');
+
+        filterEl.append(feFloodEl, feCompositeEl);
+        svgFiltersElement.append(filterEl);
```
sure, I agree it's more legible

nchevobbe commented:
> > > I'm very newbie in terms of SVG filters. I looked at MDN but still I'm not sure I fully understand how filters work, but I'm not opposed.
> > > I was a bit afraid that `@property` doesn't work in the old ESR, can you double check that the button still works as expected in that version? I don't mind that the forced-colors version doesn't work properly there, but at least the main way of working should work.
> > > Thanks!
> > 
> > 
> > You mean current ESR (128) or the previous one (115) ? `@property` was added to 128, not sure how things look in 115 though
> 
> yeah I meant v115, [as far as I understand it's still supported](https://whattrainisitnow.com/release/?version=esr).

testing this I realized I don't even need `@property` as `flood-color` values can be any type of css color, even the result of a color function (e.g. `light-dark(red, blue)` works fine). I'll remove the `@property` rules

nchevobbe commented:
@julienw I checked on ESR 115 and everything seems to be okay, and I think I addressed all the comments you added, so this should be good for the final review :)

julienw commented:
don't worry about codecov, it's unhappy because the added code in index.js isn't exercised by the tests. I think that's OK, but if you'd like to exercise it, we would probably need to extract the code to a separate function in a different file (so that we could call it directly), and add something such as https://www.npmjs.com/package/jest-matchmedia-mock.

But again I'm not asking for this here :-)

julienw commented on src/index.js:
```diff
@@ -44,6 +44,57 @@ window.geckoProfilerPromise = new Promise(function (resolve) {
+      '--button-icon-active-color',
+    ];
+    for (const cssVariable of colors) {
+      let filterEl = svgFiltersElement.querySelector('#' + cssVariable);
```
`getElementById` doesn't exist on an element, it's only on `document`:
![image](https://github.com/user-attachments/assets/93951ac8-9951-4738-94ef-b1eb9479c87d)
(from https://dom.spec.whatwg.org/#interface-nonelementparentnode)

but you could use `document.getElementById` if you want.

nchevobbe replied:
yes, that's what I was using but I was probably confused about the error message that was there without the guards I added 
Switching to getElementById is fine

julienw commented on res/css/style.css:
```diff
@@ -39,6 +39,19 @@ body {
+    * a valid color (which means we can use `light-dark()` or any other color functions).
+    * The colors can be applied on SVG icons with `filter: url(#--var-name)` */
+    --button-icon-color: ButtonText;
+    --button-icon-hover-color: SelectedItem;
+    --button-icon-active-color: SelectedItem;
```
BTW the same than usual still happens: SelectedItem is not great for me in light mode. But out of scope here...

nchevobbe replied:
yeah, I need to reach to the accessibility team to sort this out
